### PR TITLE
Ft cpdt 210

### PR DESCRIPTION
**Pull request by mlear**
## Type of change

- [ ] Bug Fix
- [x] New Feature
- [ ] Refactor
- [ ] Project Build
- [ ] Project Configuration

## Submission Checklist

- [x] Have you followed the guidelines in our [Contributing document](https://github.com/fanappics/buoy/blob/master/CONTRIBUTING.md)?
- [x] Did you create appropriate tests?

## Description

This is the table component for Buoy.

## Related Issues / Bugs / Feature Requests

https://jira.fanatics.com/browse/CPDT-210

## Test Deployment URL

Replace branch-name with the name of your branch.

https://rawgit.com/fanappics/buoy/ft-cpdt-210/docs/index.html

## Deployment Notes

Notes regarding deployment the contained body of work.  These should note any
db migrations, etc.

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.

### CODE REVIEW

File: src/components/BTable/BTable.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+The headers array holds objects that represent columns
+
+| Key      | Required/Optional | Default | Description                              |
+| -------- | ----------------- | ------- | ---------------------------------------- |
+| name     | Required          |         | This is what will be displayed as the column header.  |
+| key      | Required          |         | This is the matching key for that column to access the corresponding data in the rows. |
+| sort     | Optional          |         | Enables sorting on that column.                   |
+| render   | Optional          |         | Holds the rendering function name if applicabled.                   |
```
Comment (kierachell) [inline src/components/BTable/BTable.md]: `applicabled`

File: src/components/BTable/BTable.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        html += `<span class="ion-edit" style="color: #4d90fe; font-size: 18px"></span><a href=${ data.id } class='link-group'> Edit</a>`
+        break;
+    }
+  })
+  return html
+}
+
+const headers = [
+  {name: 'Header 1', key: 'header_one', sort: true}, 
+  {name: 'Header 2', key: 'header_two', render: makeLink, sort: true},
```
Comment (kierachell) [inline src/components/BTable/BTable.md]: it's sortable, but all values are the same :(

File: src/components/BTable/BTable.vue [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        })
+        rows.push(row)
+      })
+      return rows
+    },
+    /**
+    * @param {String} sortBy
+    * Populates an object to track sort state and emits
+    * that object to the 'sort' event which a parent
+    * component can listen to using '.$on'
```
Comment (kierachell) [inline src/components/BTable/BTable.vue]: - Can't get it to sort
sortme

- also the whole header/table shifts on click

File: src/components/BTable/BTable.vue [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+      this.$forceUpdate()
+    },
+    /**
+    * @param {String} page
+    * Updates the current row the table is on, and emits
+    * an object holding the new page and the rowzsize
+    * to the 'paginate' event which a parent
+    * component can listen to using '.$on'
+    */
+    onPaginateClick: function(page) {
```
Comment (kierachell) [inline src/components/BTable/BTable.vue]: I changed tableData to:
```
let tableData = []
for (i=0; i < 100; i++) {
  tableData.push({
    header_one: "column " + i, 
    header_two: {name: "column " + i*i, id: i+1}, 
    header_three: {source:'https://raw.githubusercontent.com/fanappics/buoy/master/buoylogo.png', name: 'buoy logo'}, 
    header_four: 'column 4', 
    header_five: 'column a', 
    header_six: 'column c', 
    name: "example row " + i, 
    actions: [(i%2 ? 'delete' : 'edit')]
  })
} 
```

and it did not paginate:

1000rows

File: src/components/BTable/BTable.vue [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    },
+    /**
+    * The total number of rows for the table.
+    */
+    totalResults: {
+      type: Number,
+      required: true
+    },
+    /**
+    * The different row sizes availablt to the user, will default to [10, 20, 50] if not provided.
```
Comment (kierachell) [inline src/components/BTable/BTable.vue]: `available` is hates you... I hate it too

File: src/components/BTable/BTable.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  {
+    header_one: 'column 1', 
+    header_two: {name: 'column 2', id: 2}, 
+    header_three: {source:'https://raw.githubusercontent.com/fanappics/buoy/master/buoylogo.png', name: 'buoy logo'}, 
+    header_four: 'column 4', 
+    header_five: 'column 5', 
+    header_six: 'column 6', 
+    name: 'example row 1', 
+    actions: ['delete']
+  },{header_one: 'column 1', header_two: {name: 'column 2', id: 2}, header_three: {source:'https://raw.githubusercontent.com/fanappics/buoy/master/buoylogo.png', name: 'buoy logo'}, header_four: 'column 4', header_five: 'column 5', header_six: 'column 6', name: 'example row 2', actions: ['delete', 'edit']},
```
Comment (kierachell) [inline src/components/BTable/BTable.md]: what does `id: 2` do?

File: src/components/BTable/BTable.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+const makeImage = function(data) {
+  return `<img src="${ data.source }" alt="${ data.name }" width="30" height="30">`
+}
+
+const makeLinkGroup = function(data) {
+  let html = ''
+  data.forEach(function(action) {
+    switch(action) {
+      case 'delete':
+        html += `<span class="ion-ios-trash" style="color: #4d90fe; font-size: 18px"></span><a href=${ data.id } class='link-group'> Delete</a>`
```
Comment (kierachell) [inline src/components/BTable/BTable.md]: maybe change these to `alert` or something so that it doesn't 404 when someone clicks...

or use the popup?

**Review by ghost — DISMISSED** [review]
a few nits

File: src/components/BTable/BTable.md (hunk 2 above)
Comment (mlear) [reply src/components/BTable/BTable.md]: im not sure what you mean?  This just lets the column be sorted, the sorting will be handled by the end user.

File: src/components/BTable/BTable.md (hunk 1 above)
Comment (mlear) [reply src/components/BTable/BTable.md]: 👍

File: src/components/BTable/BTable.vue (hunk 3 above)
Comment (mlear) [reply src/components/BTable/BTable.vue]: what are you testing this on?  It should't sort, sorting will be handled by the user.

File: src/components/BTable/BTable.vue (hunk 4 above)
Comment (mlear) [reply src/components/BTable/BTable.vue]: it won't paginate like that.  Those arrows fire off events for the user to hook into.  Like sorting, pagination is handled by the user.

File: src/components/BTable/BTable.vue (hunk 5 above)
Comment (mlear) [reply src/components/BTable/BTable.vue]: my typing is hurible

File: src/components/BTable/BTable.md (hunk 6 above)
Comment (mlear) [reply src/components/BTable/BTable.md]: inserts the id into the link, all functionality that would be handled by the user

File: src/components/BTable/BTable.md (hunk 7 above)
Comment (mlear) [reply src/components/BTable/BTable.md]: the popup didnt' exist when I started building this.  I will integrate it though when its done

**Review by ghost — APPROVED** [review]

File: src/components/BTable/BTable.vue (hunk 4 above)
Comment (kierachell) [reply src/components/BTable/BTable.vue]: gotcha

**Comment by mlear** [comment]
the up/down arrow combination indicates the header is sortable and currently unsorted.  Which browser are you using where you see the uncentered image?  The select button is odd as well

File: src/components/BTable/BTable.vue (hunk 4 above)
Comment (eleanormal) [reply src/components/BTable/BTable.vue]: @mlear Is that documented in the styleguide?

Comment (mlear) [reply src/components/BTable/BTable.vue]: its documented in the code for the methods, but currently we are not including the methods in the styleguide... I'll update the markdown to mention this specifically

File: docs/main.css [hunk 8]
```diff
@@ -222,6 +222,13 @@ textarea.b-error-border,
...
   opacity: 0.4;
 }
+a:link {
+  color: #4d90fe;
+  text-decoration: none;
+}
+.link-group {
+  margin-right: 0.75rem;
+}
 li {
```
Comment (ghost-) [inline docs/main.css]: These will affect everything in the document, not jut buoy components.  Do we want to use a class or something else to differentiate so we don't override the browsers default styling?

File: src/main.styl [hunk 9]
```diff
@@ -84,5 +84,12 @@ input.b-error-border, textarea.b-error-border, .b-input.b-error-border
...
   opacity 0.4
 
+a:link
+  color $cornflower-blue
+  text-decoration none
+  
+.link-group
+  margin-right .75rem
+
 li
```
Comment (ghost-) [inline src/main.styl]: See above comment.

**Review by AnalogGhost — APPROVED** [review]
Just a few questions about the global styling.  Nothing to hold up the PR though.

File: docs/main.css (hunk 8 above)
Comment (mlear) [reply docs/main.css]: thats totally in the wrong place, Its an artifact of a refactor, I'll fix!

**Pull request merged**